### DISCUSSION TITLE
Does UltimateKEYS work with UK ISO keyboards?

### DISCUSSION THREAD

**Discussion by liegepr** [pieter-degroote/UltimateKEYS#16] [General]
Hi,
I am confused with the keyboard layout image showing an additional key right to L-shift as compared with the ANSI layout. Does that mean that this key, being also a compose key, is intended to be used with UK ISO keyboards?

**Comment by pieter-degroote** [comment]
This is right. I have 'mirrored' the Compose Key (from the top left position) to the ISO key, making the Compose Key somewhat better and more easily accessible.
In case you use a UK keyboard, you 'are favoured with' that additional Compose Key.
That way, I have simply _defined_ what the ISO key should do, in case it is present there.
Please note that the keyboard layout image is more of a 'hybrid blend' of a 104-key US keyboard, where I just put a 105th ISO key there as an illustration.
When applying this layout to a physical UK keyboard, the following characters will correspond to their US-based positions: `"`, `@`, `#` and `~`.

On a 104-key US keyboard, everything else should just work in the same way. Since the ISO key is not present on all physical keyboards, this key does not introduce any new or exclusive features.
_I believe that 105-key US keyboards will be extremely rare to find, if they exist at all..._

(As for now, I touch-type on a (105-key) Belgian AZERTY keyboard which is configured in the UltimateKEYS layout, as this one is simply really solid/durable...)

**Reply by liegepr** [reply]
Thanks, Pieter, I will give it a try.

**Comment by liegepr** [comment]
The sorting order for letters with diacritical marks is disturbing me. The acute accent should precede the grave accent. In french for example, the sorting order for diacritic-modifed e is e, é, è, ê, ë (https://fr.wikipedia.org/wiki/Alphabet_fran%C3%A7ais). This sorting order is well in line with the European ordering rules (https://en.wikipedia.org/wiki/European_ordering_rules). Further, it is common practice to place an acute ó before a grave ò (https://en.wikipedia.org/wiki/Diacritic). I guess you are referring to unicode conventions.

**Comment by pieter-degroote** [comment]
Well, I get the point. UltimateKEYS actually started off as simply a fork of EurKEY (see https://eurkey.steffen.bruentjen.eu/), although I gradually started making changes to the main layout.

As such, for instance, I removed the ligature IJ for Dutch (_always_ I and J separately in Dutch) and put the ligature OE there, making it more consistent with 'US International' on Linux.

Further on, I found out to make the main layout more 'appropriate' for Western European languages/users by swapping out ð/Ð, þ/Þ, ý/Ý and ÿ/Ÿ, and to put the vowels with a circumflex (â, ê, î, ô, û) there instead for i.e. French and Portuguese. Also, two letters with a tilde (ã and õ) have been added.
(The former letters for Icelandic (and Faroese) are much less common than those with a circumflex or tilde.)

That way, the acute and grave accented letters are on the same places as with EurKEY. I believe that EurKEY might be slightly more optimized for German (and indeed English), since all letters with a diaeresis/umlaut correspond to their unaccented versions on the layout.

Opposite to the 'US International [with AltGr dead keys]' layout, which is more optimized for both English and Spanish (and additionally, German), since á, é, í, ó, ú and ñ correspond to their unaccented versions on this layout (most natural places for Spanish).

As for now, the main layout is mostly suitable for English, German, Dutch, French, Spanish, Italian, Portuguese, Norwegian, Swedish, Danish and some local variants. Other letters and symbols (for i.e. Eastern Europe) are accessible via the Compose Key.

Note that I 'got rid of' the dead keys on the EurKEY layout, and replaced them with that single Compose Key on the top left position (and mirrored it to the ISO key). This because I concluded that it mostly does not make sense anymore to have a 'dead acute', 'dead grave', 'dead diaeresis', 'dead circumflex', 'dead tilde'..., since their common occurrences are already on the main layout, making those separate dead keys mostly useless.

You can easily compare both keyboard layout images:

UltimateKEYS: https://pieter-degroote.github.io/UltimateKEYS/images/UltimateKEYS%20-%20Keyboard%20Layout%20Image.png

EurKEY: https://eurkey.steffen.bruentjen.eu/pics/layout.png (https://eurkey.steffen.bruentjen.eu/)

**Comment by pieter-degroote** [comment]
As for now, you can also find a conceptual version in the 'Lab Tests and Concepts', featuring a 'regular' dead key on the grave accent and ISO key.

This dead key can be used as a substitute for holding down the Right Alt. Pressing it twice results in an empty grave accent.
(Note that there is a minor difference when the Caps Lock is enabled. The letters ó and ò can only be capitalized via `<grave/iso> <less>` and `<grave/iso> <greater>`, regardless of the Caps Lock state.)

UltimateKEYS - Grave Accent: https://github.com/pieter-degroote/lab-tests-concepts/tree/master/UltimateKEYS%20-%20Grave%20Accent%20mod

**Comment by liegepr** [comment]
Hi Pieter,
Thanks for your detailed answers. Indeed, Eurkey was initiated in Germany and is best optimised for writing German text. It can be used, of course, for other languages, but with some limitations with respect to ergonomics. The worst case is probably the French "e" because this letter is combined with four diacritics which are spread over the keyboard with almost no rationale behind.

That being said, it is not surprising for a “European keyboard” to have a slightly steeper learning curve when compared to a language-specific layout (fr-oss, qwerty-fr, qwerty-lafayette, etc.)

I am still of the opinion that the layout should be made compatible with the European ordering rules as far as ´ and ` are concerned (this cannot be applied for all diacritics). Having é before è would have minimal side effects and would better comply with common ordering. And the same for other relevant vowels.

I am fine with your compose key and I don't like that much layouts with numerous dead keys. It works well for diacritics in direct access. Less comfortable if these can only be keyed in using ALT-GR or ALT-GR + SHIFT, but solutions based on dead keys do not perform better. Further, the compose key allows entering very easily a Greek letter or a math symbol which is welcome when writing scientific reports. A feature that might be useful also for Julia programmers (unicode symbols being allowed in this language).

**Comment by pieter-degroote** [comment]
Nice to find out from you.

As for now, I have created a test version, where those accented letters have switched places.

See in the 'Lab Tests and Concepts': https://github.com/pieter-degroote/lab-tests-concepts/tree/master/UltimateKEYS%20-%20Latest
(*updated link*)

**Keyboard layout image:**

![UltimateKEYS - Keyboard Layout Image](https://raw.githubusercontent.com/pieter-degroote/lab-tests-concepts/master/UltimateKEYS%20-%20Latest/UltimateKEYS%20-%20Keyboard%20Layout%20Image.png)

At the moment, I also find it slightly more natural to have the acute accents on the left and the grave accents on the right...

As for now, I doubt a little whether I should copy these changes into the main version, since this is more of a 'major overhaul'. I hope that you enjoy it.

Edit: The 'reported bug' of MSKLC/Windows has been removed, since it only occurred on an obsolete version of Windows 10 (in VirtualBox, without network or internet). The latest version of Windows 10 does not seem to be affected.

**Comment by pieter-degroote** [comment]
**As for now, I have also created an issue for this topic: https://github.com/pieter-degroote/UltimateKEYS/issues/17**

**Comment by liegepr** [comment]
It would be interesting to receive some feedback from other users before any merging decision.

**Comment by pieter-degroote** [comment]
Actually, in that case, I am considering to put the current 'main' release into the 'legacy' repository...
That way, users still have the choice to go back, in case they prefer the former version over the latter.

At this time, the latest legacy release is 2.5 years old, having a great number of _redundant_ dead key combinations,... (in my opinion). That means that there would finally be a new release in this repository.

**Comment by pieter-degroote** [comment]
During the past week, I quickly surveyed my friends and family by showing them both the images.
Although they don't use this layout (mostly AZERTY instead in Belgium), the greater majority has a slight or appreciable/significant preference for the ordering "á-à, é-è, í-ì, ó-ò, ú-ù" (over the current placements in EurKEY and this (former) main repository).

I consider to take this into account for UltimateKEYS's future development.

At the time, I am also testing the new pilot version and I experience it as 'slightly more natural'.

Edit: The 'reported bug' of MSKLC/Windows has been removed, since it only occurred on an obsolete version of Windows 10 (in VirtualBox, without network or internet). The latest version of Windows 10 does not seem to be affected.

**Comment by pieter-degroote** [comment]
Now I have created a test version, having the accented letters swapped.
Further on, the bullet (•) now got its place where the ellipsis (…) was, next to the right Shift (the ellipsis is usually typed as three separate dots).

See in the 'Lab Tests and Concepts': https://github.com/pieter-degroote/lab-tests-concepts/tree/master/UltimateKEYS%20-%20Latest

**Keyboard layout image:**

![UltimateKEYS - Keyboard Layout Image](https://raw.githubusercontent.com/pieter-degroote/lab-tests-concepts/master/UltimateKEYS%20-%20Latest/UltimateKEYS%20-%20Keyboard%20Layout%20Image.png)

(Note that the Unicode 'horizontal ellipsis' can be typed anyway via: Compose, dot (.), dot (.) or Compose, dot (.), 3)

**Comment by liegepr** [comment]
Thanks for this update. Never entered a bullet using my keyboard until now. The same for the ellipsis though I should do it regularly when typing French. Not a fan of daggers and double daggers which look awful in most font designs. They are mainly used in footnotes, but a second reason for me to avoid them is their double meaning (obsolete/dead). Quite a few scientific journals have it in their guide to manuscript writing, however.

**Comment by pieter-degroote** [comment]
Well, I made a few practical and theoretical reasonings when putting some characters, like §, †, ‡ on the layout, since they were also included in the former ANSI code pages (like Windows: Western, and Windows: Central Europe)... That way, I believe it might be better to put some more 'universal' characters on the main layout, rather than more 'exotic' ones (like →, ↓, ↔).
I have not considered Unicode fractions (½, ¼, ¾, etc.), since they are far to limited, in my opinion.
In the longer past, the copyright sign (©) was also on the main layout, but some people, including me, found it less appropriate, since this project is entirely open source (GPL v3)...

**Comment by pieter-degroote** [comment]
Since I also agree that the dagger (†) and double dagger (‡) are not that 'commonplace', I have now put two music notes onto the main layout (♪ and ♫). Those specific 'eighth note' (♪) and 'beamed eighth notes' (♫) even already existed back in the 'olden' DOS days (then the ASCII code points were 0x0D and 0x0E respectively...).
I found out that those specific two characters are often used to mention that something is a melody, or is being sung. They are also defined in most of the available fonts.

Further on, they have been put on the main layout in the same order as the single (apostrophe) and double quotation marks, making they 'match' with each of them.

**Keyboard layout image:**

![UltimateKEYS - Keyboard Layout Image](https://raw.githubusercontent.com/pieter-degroote/lab-tests-concepts/master/UltimateKEYS%20-%20Latest/UltimateKEYS%20-%20Keyboard%20Layout%20Image.png)

**Reply by pieter-degroote** [reply]
As for now, these changes have been merged into the latest release:

https://github.com/pieter-degroote/UltimateKEYS/releases&nbsp;&nbsp; [![GitHub Release](https://img.shields.io/github/release/pieter-degroote/UltimateKEYS.svg)](https://github.com/pieter-degroote/UltimateKEYS/releases)

The former version (+ minor update) has now been moved to the [legacy repository](https://github.com/pieter-degroote/UltimateKEYS-legacy). :)

**Comment by pieter-degroote** [comment]
During last months, I also presented the UltimateKEYS main (or base) layout as a candidate for inclusion into 'xkeyboard-config' (X11 - Linux).

See also: https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/issues/498

Closed merge request - Not accepted (yet), since it is 'currently too niche':
https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/merge_requests/779

UltimateKEYS-core (main/base layout): https://gitlab.freedesktop.org/pieter-degroote/UltimateKEYS-core

**Comment by pieter-degroote** [comment]
***=\> Update&nbsp;: &nbsp;Please see the latest release&nbsp;:***

**UltimateKEYS&nbsp;:**&ensp;&emsp;&emsp;&emsp;[![UltimateKEYS Release](https://img.shields.io/github/release/pieter-degroote/UltimateKEYS.svg)](https://github.com/pieter-degroote/UltimateKEYS/releases)
**UltimateKEYS Core&nbsp;:**&emsp;[![Core Release](https://img.shields.io/github/release/pieter-degroote/UltimateKEYS-core.svg?label=core)](https://github.com/pieter-degroote/UltimateKEYS-core/releases)
